cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt

```diff
@@ -0,0 +1,68 @@
+package com.dropbox.android.external.cache4
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.Lock
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+/**
+ * Provides a mechanism for performing key-based synchronization.
+ */
+internal class KeyedSynchronizer<Key : Any> {
+
+    private val keyBasedLocks: MutableMap<Key, LockEntry> = hashMapOf()
+
+    private val mapLock = Any()
+
+    /**
+     * Executes the given [action] under a lock associated with the [key].
+     * When called concurrently, all actions are executed sequentially for the same [key].
```

Comment:
sequentially may imply FIFO or fairness that we do not intend to guarantee. lets stick with a more neutral language

```suggestion
     * When called concurrently, all actions associated with the same [key] are mutually exclusive.
```